Add GEFS support to rocoto XML generation

WalterKolczynski-NOAA opened the pull request:
**Description**

Adds GEFS support to rocoto XML generation.

To facilitate this, a few classes were refactored into abstract classes with child classes for each of the three `$NET`/`$MODE` combinations (gfs/cycled, gfs/forecast-only, and gefs). These child classes then implement the details particular to their configuration. A generic factory class from @aerorahul was added to pygw and is used to generate objects for the appropriate class based on `$NET` and `$MODE`.

GEFS is implemented as only forecast jobs right now: a fcst job intended for the control and a metatask of efcs jobs intended for the perturbed members. A couple of additional config files were needed in order to parse the configs; these were copied from the gfs versions. Beyond that, changes have not yet been made to update any configs for GEFS. For instance, in contrast to GFS, even the control member of GEFS will want `MEMDIR` defined (to 'mem000').

Also resolves all shellcheck issues in `config.fcst` and `config.nsst`.

Fixes #1577

**Type of change**

- [x] New feature (non-breaking change which adds functionality)

**How Has This Been Tested?**

- [x] GFS cycled experiment produces identical workflow XMLs
- [x] GFS forecast-only experiment produces identical workflow XMLs
- [x] GEFS experiment produces a workflow XML that looks correct on visual inspection
- [x] GFS cycled on Orion
- [x] GFS P8 on Orion

**Checklist**

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [x] New and existing tests pass with my changes

WalterKolczynski-NOAA committed:
Add GEFS support to rocoto XML generation

Adds GEFS support to rocoto XML generation.

To facilitate this, a few classes were refactored into abstract
classes with child classes for each of the three `$NET`/`$MODE`
combinations (gfs/cycled, gfs/forecast-only, and gefs). These child
classes then implement the details particular to their configuration.

GEFS is implemented as only forecast jobs right now: a fcst job
intended for the control and a metatask of efcs jobs intended for the
perturbed members. A couple of additional config files were needed in
order to parse the configs; these were copied from the gfs versions.
Beyond that, changes have not yet been made to update any configs for
GEFS. For instance, in contrast to GFS, even the control member of
GEFS will want `MEMDIR` defined (to 'mem000').

Fixes #1577

WalterKolczynski-NOAA committed:
Remove unused import in setup_xml

Added a pprint at some point during debugging and forgot to remove
the import when I was done.

Refs: #1577

WalterKolczynski-NOAA committed:
Bring fcst and nsst configs into shellcheck compliance

Updates `config.fcst` and `config.nsst` to address all shellcheck
warnings.

WalterKolczynski-NOAA committed:
Update workflow factory names

Update the workflow generation factory names to be more explicit
about what kind of factory it is.

Refs: #1577

WalterKolczynski-NOAA committed:
Remove bit of sfc cycling setting from gefs config.base

`config.fcst` isn't being trimmed down for GEFS yet, but this bit
was causing shellcheck to complain. There is no reason to add a
shellcheck directive if it is about to be removed anyway.

Refs: #1577

WalterKolczynski-NOAA committed:
Fix minor bugs from shellcheck fixes

WalterKolczynski-NOAA commented:
With all the config modifications now to fix all the shellcheck warnings, I'm actually running the forecast.

XianwuXue-NOAA commented on parm/config/gefs/config.fcst:
```diff
@@ -0,0 +1,366 @@
+   export rf_cutoff=1.0e3
+   export d2_bg_k1=0.20
+   export d2_bg_k2=0.0
+fi
```
Does GEFS need this?

XianwuXue-NOAA replied:
I mean lines 144-149

WalterKolczynski-NOAA replied:
These are just copied from GFS for now. They will be pruned later.

XianwuXue-NOAA commented on parm/config/gefs/config.fcst:
```diff
@@ -0,0 +1,366 @@
+export DO_SKEB=${DO_SKEB:-"NO"}
+export DO_SHUM=${DO_SHUM:-"NO"}
+export DO_LAND_PERT=${DO_LAND_PERT:-"NO"}
+export DO_CA=${DO_CA:-"YES"}
```
For mem000, DO_CA should be NO or YES?

WalterKolczynski-NOAA replied:
See above, these are just copied from GFS for now. Future PR can update them (so we can see the changes).

When we get there, set cellular automata to whatever v12 is using unless/until we get guidance from MDAB.

XianwuXue-NOAA commented on parm/config/gefs/config.fcst:
```diff
@@ -0,0 +1,366 @@
+# Disable the use of coupler.res; get model start time from model_configure
+export USE_COUPLER_RES="NO"
+
+if [[ "${CDUMP}" =~ "gdas" ]] ; then # GDAS cycle specific parameters
```
Can CDUMP=gefs? For the conditions Lines 280-349, it should keep only for GEFS part. Or I will do this.

WalterKolczynski-NOAA replied:
Same as above. We'll strip it down later.

XianwuXue-NOAA commented on parm/config/gfs/config.fcst:
```diff
@@ -276,10 +277,10 @@ export deflate_level=1
 export USE_COUPLER_RES="NO"
 
-if [[ "$CDUMP" =~ "gdas" ]] ; then # GDAS cycle specific parameters
+if [[ "${CDUMP}" =~ "gdas" ]] ; then # GDAS cycle specific parameters
```
Same things for these conditions. It should only keep gefs parts

WalterKolczynski-NOAA replied:
This is the GFS one.

XianwuXue-NOAA reviewed (commented):
I am still working on the tests, just post some comments

WalterKolczynski-NOAA commented:
> I am still working on the tests, just post some comments

Note this probably does not work to actually run the GEFS forecast. But you should get a workflow with just a fcst task (for the control) and an efcs metatask (for the perturbed members)

XianwuXue-NOAA commented:
> > I am still working on the tests, just post some comments
> 
> Note this probably does not work to actually run the GEFS forecast. But you should get a workflow with just a fcst task (for the control) and an efcs metatask (for the perturbed members)

Why uses fcst task instead of efcs for all members since all members should use the same settings, including mem000.

WalterKolczynski-NOAA commented:
> > > I am still working on the tests, just post some comments
> > 
> > 
> > Note this probably does not work to actually run the GEFS forecast. But you should get a workflow with just a fcst task (for the control) and an efcs metatask (for the perturbed members)
> 
> Why uses fcst task instead of efcs for all members since all members should use the same settings, including mem000.

Because it doesn't. The control doesn't use stochastic physics, for one. Plus, it will be easier from a code management standpoint because it means the same thing as in the GFS.

aerorahul commented on workflow/rocoto/workflow_tasks.py:
```diff
@@ -189,6 +184,41 @@ def get_task(self, task_name, *args, **kwargs):
+        -------
+        Tasks: New object of the appropriate type for the net.
+        '''
+        if app_config.net in ['gfs']:
```
I think this should work:
```python
me = sys.modules[__name__]
class_name = f"{app_config.net.upper()Tasks}"
try:
    return getattr(me, class_name)(app_config, cdump)
except AttributeError:
    raise AttributeError(f"{class_name} is not a valid class in {me}"
```

It would help you everywhere you have `if-else` conditions.

WalterKolczynski-NOAA replied:
I think the better approach would be to just change it to the uwtools Factory. Not sure if we should do that now, or get this in so things can move forward and then refactor the factories later.

aerorahul reviewed (commented):
looks good.
A suggestion if you want to consider making the factory truly agnostic.

WalterKolczynski-NOAA committed:
Add generic factory to pygw and update workflow factories

Adds a generic factory class to pygw. This was originally created
by @aerorahul and is added here with only minor changes to update
the package name and add additional documentation.

Updated the factories in workflow generation to use the new factory
class.

WalterKolczynski-NOAA committed:
Reorder cycledefs

WalterKolczynski-NOAA committed:
Fix line spacing in workflow_xml

WalterKolczynski-NOAA committed:
Convert test_factory indentation to spaces

aerorahul commented on workflow/rocoto/workflow_tasks.py:
```diff
@@ -27,6 +28,8 @@ class Tasks:
                    'waveawipsbulls', 'waveawipsgridded', 'wavegempak', 'waveinit',
                    'wavepostbndpnt', 'wavepostbndpntbll', 'wavepostpnt', 'wavepostsbs', 'waveprep']
 
+    tasks_factory = Factory(__qualname__)
```
I think you want this out of the `class Tasks` rather than a class attribute.

aerorahul replied:
Are we going to use `Factory` the way it was intended to be used or keep using it this way?

aerorahul commented on workflow/rocoto/workflow_tasks.py:
```diff
@@ -1417,6 +1426,40 @@ def earc(self):
+        return task
+
+
+Tasks.tasks_factory.register('gfs', GFSTasks)
+Tasks.tasks_factory.register('gefs', GEFSTasks)
```
This is not the same as:
`tasks_factory.register('gfs', GFSTasks)` as it should be.
When you call `Tasks.tasks_factory...`, you are spinning up another instance of `Tasks`. The memory address of this instance will be different than the one for `gefs`.
I don't know how this is working.

WalterKolczynski-NOAA replied:
`tasks_factory` is a static field of Tasks. It doesn't create a new Tasks object at all. A single factory is created for Tasks at class definition time, which is what is accessed here.

aerorahul replied:
I see that, and see my previous comment.  `tasks_factory` cannot be a member of the class `Tasks` as `tasks_factory` is and should be an instance of `class TasksFactory`, which is inherited from `Factory`.  Such that:
```python
class TasksFactory(Factory):
```
and 
```python
tasks_factory = TasksFactory()
tasks_factory.register('gfs', GFSTasks)
```

aerorahul replied:
I would like to see it used in the way it was supposed/designed, rather than using it any other way because it "works"

aerorahul commented on workflow/rocoto/workflow_tasks.py:
```diff
@@ -1456,7 +1499,7 @@ def get_wf_tasks(app_config: AppConfig) -> List:
     # Loop over all keys of cycles (CDUMP)
     for cdump, cdump_tasks in app_config.task_names.items():
-        task_obj = Tasks(app_config, cdump)  # create Task object based on cdump
+        task_obj = Tasks.tasks_factory.create(app_config.net, app_config, cdump)  # create Task object based on cdump
```
This should also be `tasks_factory.create` and not `Tasks.tasks_factory`.

aerorahul reviewed (changes requested):
Sorry I started reviewing in the morning but got distracted.

aerorahul commented on ush/python/pygw/src/pygw/factory.py:
```diff
@@ -0,0 +1,134 @@
+        self._name = f'{name}Factory'
+        self._builders = {}
+
+        # Register {name}Factory as importable from uwtools.factory
```
```suggestion
        # Register {name}Factory as importable from pygw.factory
```

aerorahul reviewed (changes requested):
still waiting on requested updates + one correction.

WalterKolczynski-NOAA commented:
I'm working on the factory stuff. Need to make a new one for all three uses. Also splitting out the new child classes into separate files as we discussed.

WalterKolczynski-NOAA committed:
Refactor factory use

Creates factory subclasses for each factory and use global variables
to hold the factory instantiation rather than static fields in the
class the factory is for.

Also separates the child classes added in this PR into individual
files. `get_wf_tasks()` had to be separated from `Tasks` and
`create_wf_task()` to avoid circular imports.

Refs: #1577

WalterKolczynski-NOAA committed:
Convert tab indent to spaces

A couple of the new python files wound up with tab indents instead
of spaces.

WalterKolczynski-NOAA commented:
@TerrenceMcGuinness-NOAA I have no indication what failed about experiment setup. I can't even find a log in the CI directory where that information would be held.

TerrenceMcGuinness-NOAA commented:
Here is the EXPDIR:` /scratch1/NCEPDEV/global/Terry.McGuinness/GFS_CI_ROOT/PR/1601/RUNTESTS/C96_atm3DVar/EXPDIR/C96_atm3DVar` Yes I see there is no error reporting or logging on the `setup_exp.py` fail.  Let's put in a PR request on this for sure.

TerrenceMcGuinness-NOAA commented:
Let me try to reproduce the error.

TerrenceMcGuinness-NOAA commented:
Here's the error.  It was with setup_xml.py: 

```shell
(gfs_workflow) Terry.McGuinness (hfe01) workflow (feature/gefs_workflow) $ ./setup_xml.py /scratch1/NCEPDEV/global/Terry.McGuinness/GFS_CI_ROOT/PR/1601/RUNTESTS/C96_atm3DVar/EXPDIR/C96_atm3DVar
Traceback (most recent call last):
  File "/scratch1/NCEPDEV/global/Terry.McGuinness/GFS_CI_ROOT/PR/1601/global-workflow/workflow/./setup_xml.py", line 9, in <module>
    from applications.application_factory import app_config_factory
ModuleNotFoundError: No module named 'applications.application_factory'; 'applications' is not a package
(gfs_workflow) Terry.McGuinness (hfe01) workflow (feature/gefs_workflow) $ 
```

WalterKolczynski-NOAA committed:
Add applications files that were not added previously

For some reason, the new applications directory and its contents did
not appear in `git status` and thus were overlooked in being added.

WalterKolczynski-NOAA commented:
Thanks Terry. The new directory did not appear in `git status`, so I overlooked adding it to the commit. Fixed now.

WalterKolczynski-NOAA committed:
Fix PEP-8 violations

WalterKolczynski-NOAA committed:
Fix pystyle wanting two blank lines at end of file

WalterKolczynski-NOAA committed:
Fix end-of-file pystyle complaints

aerorahul commented on workflow/rocoto/gefs_xml.py:
```diff
@@ -0,0 +1,30 @@
+    def __init__(self, app_config: AppConfig, rocoto_config: Dict) -> None:
+        super().__init__(app_config, rocoto_config)
+
+    def _get_cycledefs(self):
```
This shouldn't be `_get_cycledefs` any more as it is a concrete implementation *and* no longer a private

aerorahul commented on workflow/rocoto/rocoto_xml_factory.py:
```diff
@@ -0,0 +1,16 @@
+from rocoto.gefs_xml import GEFSRocotoXML
+
+
+class RocotoXMLFactory(Factory):
+
+    def __init__(self, name: str):
+        super().__init__(name)
```
Not sure why you need this.

WalterKolczynski-NOAA replied:
Factory sub-classes have been removed.

aerorahul commented on workflow/rocoto/rocoto_xml_factory.py:
```diff
@@ -0,0 +1,16 @@
+        super().__init__(name)
+
+
+rocoto_xml_factory = RocotoXMLFactory('RocotoXML')
```
```suggestion
rocoto_xml_factory = Factory('RocotoXML')
```

WalterKolczynski-NOAA replied:
Done.

WalterKolczynski-NOAA committed:
Remove factory subclasses

The individual factory classes are not necessary, so they are removed.

WalterKolczynski-NOAA committed:
Make get_cycledefs non-private

aerorahul committed:
Update ush/python/pygw/src/pygw/factory.py

aerorahul reviewed (approved):
looks good.
thanks for adding GEFS XML support.

WalterKolczynski-NOAA committed:
Fixed method name for get_cycledefs
